File: apps/android/native/core/src/main/kotlin/org/indiafoss/companion/core/PortableActivity.kt
package org.indiafoss.companion.core

import kotlinx.serialization.Serializable

@Serializable
data class PortableActivityReference(val eventId: String, val activityId: String, val proposalId: String? = null)

@Serializable
data class ActivityResolution(val status: String, val activityId: String? = null)

/** Same resolution policy as the PWA; both implementations consume the same test cases. */
object PortableActivities {
    fun resolve(reference: PortableActivityReference, eventId: String, activities: List<Activity>): ActivityResolution {
        if (reference.eventId != eventId) return ActivityResolution("wrong-event")
        if (reference.activityId.isBlank() || reference.proposalId == "") return ActivityResolution("missing")
        val matches = if (reference.proposalId != null) {
            activities.filter { it.proposalId == reference.proposalId }
        } else {
            activities.filter { it.id == reference.activityId }
        }
        val exact = matches.filter { it.id == reference.activityId }
        val candidates = exact.ifEmpty { matches }
        if (candidates.size > 1) return ActivityResolution("ambiguous")
        return candidates.firstOrNull()?.let { ActivityResolution("matched", it.id) } ?: ActivityResolution("missing")
    }
}

File: packages/model/src/portable-activity.ts
import type { Activity } from './index.js';

/** Event-scoped reference carried by a personal-data transfer. */
export interface PortableActivityReference {
  eventId: string;
  activityId: string;
  proposalId?: string;
}

export type ActivityResolution =
  { status: 'matched'; activityId: string } | { status: 'wrong-event' | 'missing' | 'ambiguous' };

/** Never guess by title, room or time: those change when the programme changes. */
export function resolvePortableActivity(
  reference: PortableActivityReference,
  eventId: string,
  activities: Pick<Activity, 'id' | 'proposalId'>[],
): ActivityResolution {
  if (reference.eventId !== eventId) return { status: 'wrong-event' };
  if (!reference.activityId.trim() || reference.proposalId === '') return { status: 'missing' };
  const matches = reference.proposalId
    ? activities.filter((activity) => activity.proposalId === reference.proposalId)
    : activities.filter((activity) => activity.id === reference.activityId);
  const exact = matches.filter((activity) => activity.id === reference.activityId);
  const candidates = exact.length ? exact : matches;
  if (candidates.length > 1) return { status: 'ambiguous' };
  const activity = candidates[0];
  return activity ? { status: 'matched', activityId: activity.id } : { status: 'missing' };
}
